### PR TITLE
Pr covered schemes mar22

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
This is an update on the schemes including the first prototype for `CoveredScheme`.

### CODE REVIEW

File: experimental/Schemes/CoveredScheme.jl [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+  return (0,0,0)
+end
+
+
+### standard constructors 
+
+function standard_covering(X::ProjectiveScheme{CRT}) where {CRT<:AbstractAlgebra.Ring}
+  if has_attribute(X, :standard_covering) 
+    return get_attribute(X, :standard_covering)
+  end
```
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl]:
```suggestion
@attr function standard_covering(X::ProjectiveScheme{CRT}) where {CRT<:AbstractAlgebra.Ring}
```

File: experimental/Schemes/CoveredScheme.jl [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+#   print(io, glueing_graph(C))
+end
+
+function Base.in(U::Spec, C::Covering)
+  for i in 1:npatches(C)
+    U == C[i] && return true
+  end
+  for i in 1:npatches(C)
+    if haskey(affine_refinements(C), C[i])
+      V = affine_refinements(C)[C[i]]
```
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl]:
```suggestion
```

File: experimental/Schemes/CoveredScheme.jl [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+                           [y[k+1]//y[i] for k in i:j-2],
+                           [1//y[i]],
+                           [y[k]//y[i] for k in j:r],
+                           y[r+1:end]),
+                      check=false
+                     )
+      add_glueing!(result, Glueing(U[i], U[j], f, g, check=false))
+    end
+  end
+  set_attribute!(X, :standard_covering, result)
```
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl]:
```suggestion
```

File: experimental/Schemes/CoveredScheme.jl [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+  return result
+end
+
+function standard_covering(X::ProjectiveScheme{CRT}) where {CRT<:MPolyQuoLocalizedRing}
+  if has_attribute(X, :standard_covering) 
+    return get_attribute(X, :standard_covering)
+  end
```
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl]:
```suggestion
@attr function standard_covering(X::ProjectiveScheme{CRT}) where {CRT<:MPolyQuoLocalizedRing}
```

File: experimental/Schemes/CoveredScheme.jl [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+                           [1//y[i]],
+                           [y[k]//y[i] for k in j:r],
+                           y[r+1:end]),
+                      check=false
+                     )
+      add_glueing!(result, Glueing(U[i], U[j], f, g, check=false))
+    end
+  end
+  covered_projection = CoveringMorphism(result, Covering(Y), pU)
+  set_attribute!(X, :standard_covering, result)
```
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl]:
```suggestion
```

File: experimental/Schemes/CoveredScheme.jl [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+  end
+  error("covering not listed")
+end
+
+getindex(X::CoveredScheme, i::Int, j::Int) = X[X[i], X[j]]
+
+
+set_name!(X::CoveredScheme, name::String) = set_attribute!(X, :name, name)
+name_of(X::CoveredScheme) = get_attribute(X, :name)::String
+has_name(X::CoveredScheme) = has_attribute(X, :name)
```
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl:528]: I guess these could be added to AbstractAlgebra alongside `set_name!` there (of course after changing them to be for `Any`

File: experimental/Schemes/CoveredScheme.jl [hunk 7]
```diff
@@ -0,0 +1,839 @@
...
+
+function dim(X::CoveredScheme) 
+  if !has_attribute(X, :dim)
```
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl]: Should use `@attr`

File: experimental/Schemes/ProjectiveSchemes.jl [hunk 8]
```diff
@@ -725,25 +725,90 @@ function dehomogenize(
...
 
+function (f::Oscar.MPolyAnyMap{<:MPolyRing, <:MPolyQuoLocalizedRing, <:Nothing})(a::MPolyElem)
+  if !has_attribute(f, :lifted_map)
+    S = domain(f)
+    W = codomain(f)
+    L = localized_ring(W)
+    g = hom(S, L, lift.(f.img_gens))
+    set_attribute!(f, :lifted_map, g)
+  end
+  g = get_attribute(f, :lifted_map)
```
Comment (fingolfin) [inline experimental/Schemes/ProjectiveSchemes.jl:800]: Perhaps move this into `@attr function lifted_map ...` and then just call that here.

File: experimental/Schemes/CoveredScheme.jl [hunk 9]
```diff
@@ -0,0 +1,839 @@
...
+  for i in 1:length(coverings(X))
+    C == coverings(X)[i] && return i
+  end
+  error("covering not listed")
+end
+
+getindex(X::CoveredScheme, i::Int, j::Int) = X[X[i], X[j]]
+
+
+set_name!(X::CoveredScheme, name::String) = set_attribute!(X, :name, name)
```
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl]: This is redundant, the existing `set_name!` does exactly this.

File: experimental/Schemes/CoveredScheme.jl [hunk 10]
```diff
@@ -0,0 +1,839 @@
...
+function empty_covered_scheme(R::RT) where {RT<:AbstractAlgebra.Ring}
+  return CoveredScheme(empty_spec(R))
+end
+
+
+function Base.show(io::IO, X::CoveredScheme)
+  if has_name(X)
+    print(io, name_of(X))
+    return
+  end
```
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl]: With the naming system in AbstractAlgebra, one would instead do this:
```suggestion
  @show_name(io, X)
```
However, I am not sure it should be used here... It has some funky code to "guess" names and personally I am ambivalent about it...

**Review by fingolfin — COMMENTED** [review]
A few suggestions

File: experimental/Schemes/ProjectiveSchemes.jl [hunk 11]
```diff
@@ -186,11 +186,11 @@ On ``X ⊂ ℙʳ(A)`` this returns ``r``.
 fiber_dimension(P::ProjectiveScheme) = P.r
 
 @Markdown.doc """
-    homogeneous_coordinate_ring(X::ProjectiveScheme)
+    homog_poly_ring(X::ProjectiveScheme)
```
Comment (thofma) [inline experimental/Schemes/ProjectiveSchemes.jl]: Just so that I don't forget it, why the change from `homogeneous` to `homog`? Or is internal (then it should not be exported).

File: experimental/Schemes/CoveredScheme.jl [hunk 12]
```diff
@@ -0,0 +1,839 @@
...
+    end
+    for (X, Y) in keys(glueings)
+      X in patches || error("glueings are not compatible with the patches")
+      Y in patches || error("glueings are not compatible with the patches")
+      if haskey(glueings, (Y, X))
+        if check
+          inverse(glueings[(X, Y)]) == glueings[(Y, X)] || error("glueings are not inverse of each other")
+        end
+      else
+	glueings[(Y, X)] = inverse(glueings[(X, Y)])
```
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl]: Tabs -> spaces

**Comment by fingolfin** [comment]
@HechtiDerLachs you overlooked at least one `homog_poly_ring` hence tests fail
